src/Creative/InLine/Linear.php

```diff
@@ -53,22 +60,37 @@ public function setDuration($duration)
 
         return $this;
     }
-
+    
+    
+    /**
+     * Creates a DOM element if it doesn't exist already
+     *
+     * @param string $domElementName The variable that hold the DOM element
+     * @param string $tagName The name of the DOM element to be created
+     * @param string $parentTagName The name of the parent tag that the element will be appended to
+     *  if it doesn't exist
+     *
+     * @return VOID
+     */
+    private function createDomElement($domElement, $tagName, $parentTagName){
+      if (empty($this->{$domElement})) {
```

Comment:
$this->{$domElement} - it's a magic, that hard to maintain and debug.  

You need a method that just created MediaFiles and returns so i think  `getMediaFilesElement` will be good.

```
$this->getMediaFilesElement()->createMediaFile();
$this->getMediaFilesElement()->createClosedCaptionFile();
```

Reply:
The ClosedCaptionFile tag, needs to be within a ClosedCaptionFiles tag:
`<MediaFiles> <ClosedCaptionFiles> <ClosedCaptionFile>`

If I follow your idea, I would need to create a `getClosedCaptionFilesElement` method that would basically be a code clone of `getMediaFilesElement`.

like this:
`$this->getMediaFilesElement()->createMediaFile();`
`$this->getMediaFilesElement()->getClosedCaptionFilesElement()->createClosedCaptionFile();`

If I don't create the `getClosedCaptionFilesElement` method, then I would have to put the logic of getting the `<ClosedCaptionFiles>` DOM element within `createClosedCaptionFile`.

Reply:
When you copy-and-pase code that created MediaFiles element, this was code that creates same element, but writted in different places. 

Method `getClosedCaptionFilesElement` is not a clone of `getMediaFilesElement` because it needs first to create `MediaFiles` element. If code looks like similar, this does not mean thet code is similar. 

You do not need `getClosedCaptionFilesElement` because currently there is single place where this element may create - in `createClosedCaptionFile` method. 

```
public function createClosedCaptionFile() {
    if (empty($this->closedCaptionFilesDomElement) {
           $mediaFilesElement = $this->getMediaFilesElement()
            $this->closedCaptionFilesDomElement = $mediaFilesElement->getElementsByTagName('ClosedCaptionFiles')->item(0);	
            if (!$this->closedCaptionFilesDomElement) {	
                $this->closedCaptionFilesDomElement = $this->getDomElement()->ownerDocument->createElement('ClosedCaptionFiles');	
                $mediaFilesElement->appendChild($this->closedCaptionFilesDomElement);	
            }	
        }
}
```

Explicit is better than implicit.

Reply:
You suggest a method that provides the MediaFilesDomElement to `getMediaFileElement` and `closedCaptionFileDomElement` that's the idea that I had in mind.

What I don't like too much is how similarly the `CaptionFilesDomElement` is provided to `closedCaptionFileDomElement`.

   ```
 private function getMediaFilesElement(){
      if (empty($this->mediaFilesDomElement)) {
          $this->mediaFilesDomElement = $this->getDomElement()->getElementsByTagName('MediaFiles')->item(0);
          if (!$this->mediaFilesDomElement) {
              $this->mediaFilesDomElement = $this->getDomElement()->ownerDocument->createElement('MediaFiles');
              $this->getDomElement()
                  ->getElementsByTagName('Linear')
                  ->item(0)
                  ->appendChild($this->mediaFilesDomElement);
          }
      }
      ...
    }
```

```
public function createClosedCaptionFile() 
    {
        //ensure closedCaptionFilesDomElement existence
        if (empty($this->closedCaptionFilesDomElement)) {
           $mediaFilesElement = $this->getMediaFilesElement();
            $this->closedCaptionFilesDomElement = $mediaFilesElement->getElementsByTagName('ClosedCaptionFiles')->item(0);	
            if (!$this->closedCaptionFilesDomElement) {	
                $this->closedCaptionFilesDomElement = $this->getDomElement()->ownerDocument->createElement('ClosedCaptionFiles');	
                $mediaFilesElement->appendChild($this->closedCaptionFilesDomElement);	
            }	
        }
        ...
    }
```
This is very similar code. But I don't mind.